src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -37,190 +37,103 @@
 
-  return p;
+jlong JvmtiTagMapTable::find(oop obj) {
+   //if (obj->fast_no_hash_check()) {
+   //  return 0;
+   //} else {
+     JvmtiTagMapEntry jtme(obj);
+     jlong* found = _rrht_table.get(jtme);
+     return found == NULL ? 0 : *found;
+   //}
+ }
+
+bool JvmtiTagMapTable::add(oop obj, jlong tag) {
+  JvmtiTagMapEntry new_entry(obj);
+  bool is_added = _rrht_table.put(new_entry, tag);
+  assert(is_added, "should be added");
+  return is_added;
 }
 
-void JvmtiTagMapTable::remove(oop obj) {
-  unsigned int hash = compute_hash(obj);
-  int index = hash_to_index(hash);
-  JvmtiTagMapEntry** p = bucket_addr(index);
-  JvmtiTagMapEntry* entry = bucket(index);
-  while (entry != NULL) {
-    oop target = entry->object_no_keepalive();
-    if (target != NULL && target == obj) {
-      log_trace(jvmti, table)("JvmtiTagMap entry removed for index %d", index);
-      *p = entry->next();
-      free_entry(entry);
-      return; // done
-    }
-    // get next entry and address
-    p = entry->next_addr();
-    entry = entry->next();
-  }
+bool JvmtiTagMapTable::remove(oop obj) {
+  JvmtiTagMapEntry jtme(obj);
+  return _rrht_table.remove(jtme);
 }
 
 void JvmtiTagMapTable::entry_iterate(JvmtiTagMapEntryClosure* closure) {
-  for (int i = 0; i < table_size(); ++i) {
-    for (JvmtiTagMapEntry* p = bucket(i); p != NULL; p = p->next()) {
-      closure->do_entry(p);
-    }
-  }
+  _rrht_table.iterate(closure);
 }
 
-const int _resize_load_trigger = 5;       // load factor that will trigger the resize
-static bool _resizable = true;
-
 void JvmtiTagMapTable::resize_if_needed() {
-  if (_resizable && number_of_entries() > (_resize_load_trigger*table_size())) {
-    int desired_size = calculate_resize(true);
-    if (desired_size == table_size()) {
-      _resizable = false; // hit max
-    } else {
-      if (!resize(desired_size)) {
-        // Something went wrong, turn resizing off
-        _resizable = false;
-      }
-      log_info(jvmti, table) ("JvmtiTagMap table resized to %d", table_size());
-    }
-  }
+  _rrht_table.maybe_grow();
 }
 
-// Serially remove entries for dead oops from the table and store dead oops'
-// tag in objects array if provided.
 void JvmtiTagMapTable::remove_dead_entries(GrowableArray<jlong>* objects) {
-  int oops_removed = 0;
-  int oops_counted = 0;
-  for (int i = 0; i < table_size(); ++i) {
-    JvmtiTagMapEntry** p = bucket_addr(i);
-    JvmtiTagMapEntry* entry = bucket(i);
-    while (entry != NULL) {
-      oops_counted++;
-      oop l = entry->object_no_keepalive();
-      if (l != NULL) {
-        p = entry->next_addr();
-      } else {
-        // Entry has been removed.
-        oops_removed++;
-        log_trace(jvmti, table)("JvmtiTagMap entry removed for index %d", i);
-        jlong tag = entry->tag();
-        *p = entry->next();
-        free_entry(entry);
-
-        // collect object tags for posting JVMTI events later
-        if (objects != NULL) {
-          objects->append(tag);
+  struct IsDead{
```

Comment:
Nit: space before {

Same query about using a local struct for this.

Reply:
Alternative for struct?

Reply:
Not sure ... didn't we start using C++ lambda's for some of these "closure" operations? @coleenp what is the usual pattern we use for this kind of thing?

Reply:
The `unlink` method of `ResourceHashTable` gets an ITER type and calls its `do_entry(Key&,Value&)` method. 
If we want to use lambdas, `unlink` should call the input directly and not one of its methods.

Reply:
We don't have a lambda version of unlink (yet).  The local struct is fine IMO.